### PR TITLE
[mlir][flang][openmp] Rework parallel reduction operations

### PR DESCRIPTION
**Pull request by DavidTruby**
This patch reworks the way that parallel reduction operations function to better match the expected semantics from the OpenMP specification. Previously specific omp.reduction operations were used inside the region, meaning that the reduction only applied when the correct operation was used, whereas the specification states that any change to the variable inside the region should be taken into account for the reduction.

The new semantics create a private reduction variable as a block argument which should be used normally for all operations on that variable in the region; this private variable is then combined with the others into the shared variable. This way no special omp.reduction operations are needed inside the region.

This patch only makes the change for the `parallel` operation, the change for the `wsloop` operation will be in a separate patch.

### CODE REVIEW

**Comment by tblah** [comment]
Please could you rebase this on top of the current HEAD. There has been quite a bit of churn in OpenMP.cpp.

**Comment by DavidTruby** [comment]
> Please could you rebase this on top of the current HEAD. There has been quite a bit of churn in OpenMP.cpp.

Sorry, this rebase took longer than I expected; it should be updated now.

File: flang/lib/Lower/OpenMP.cpp [hunk 1]
```diff
@@ -3483,10 +3514,8 @@ genOMP(Fortran::lower::AbstractConverter &converter,
     break;
   }
 
-  if (singleDirective) {
-    genOpenMPReduction(converter, beginClauseList);
+  if (singleDirective)
```
[the remark this answers is not in the document]
Comment (DavidTruby) [reply flang/lib/Lower/OpenMP.cpp:3521]: If genOpenMPReduction is called her it replaces the normal operations (e.g. `arith.addi`) with an `omp.reduction` operation that is no longer used/needed with the new reduction lowering. At some point when all operations have moved to the new reduction style, those operations should be removed entirely from the OpenMP dialect.

[the remark this answers is not in the document]
Comment (kiranchandramohan) [reply flang/lib/Lower/OpenMP.cpp:3521]: However we have not changed processing for the `single` directive. So it should continue to use the old flow? Although, the lack of testing suggests it was never used.

[the remark this answers is not in the document]
Comment (DavidTruby) [reply flang/lib/Lower/OpenMP.cpp:3521]: `singleDirective=true` doesn't appear to mean the `single` directive is the one being processed, but rather that _a single_ directive is being processed as opposed to a nested one. The variable name is quite unclear!

[the remark this answers is not in the document]
Comment (DavidTruby) [reply flang/lib/Lower/OpenMP.cpp:3521]: Regardless, of the directives processed by this function, only `parallel` and `teams` can actually have reduction clauses on them, and I don't think Flang supports `teams` reductions yet?

[the remark this answers is not in the document]
Comment (kiranchandramohan) [reply flang/lib/Lower/OpenMP.cpp:3521]: Yes, teams is not supported.

**Review by kiranchandramohan — APPROVED** [review]
LG.

**Review by skatrak — APPROVED** [review]
LGTM.

**Comment by jsjodin** [comment]
Could a new test be added that would give incorrect result with the old implementation but correct with this patch? It would be nice to lock in the new behavior.

**Comment by kiranchandramohan** [comment]
> Could a new test be added that would show the new capabilities?

I have added a new test test/Lower/OpenMP/parallel-reduction.f90 (https://github.com/llvm/llvm-project/pull/79308/files#diff-a659fd2e2c1970fcce5baa536110c82fbc58e30c522b8e36c769cd6eb8383436). The test should print the number of threads. Previously this test case would not work correctly since it always expected the reduction operation to be present in the region. Now we have removed this restriction allowing more cases to be handled and to match the spec better.

If there are no concerns, I would like to rebase and submit this.

**Pull request merged**